%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\PassOptionsToPackage{warn}{textcomp}
\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
% support both utf8 and utf8x syntaxes
  \ifdefined\DeclareUnicodeCharacterAsOptional
    \def\sphinxDUC#1{\DeclareUnicodeCharacter{"#1}}
  \else
    \let\sphinxDUC\DeclareUnicodeCharacter
  \fi
  \sphinxDUC{00A0}{\nobreakspace}
  \sphinxDUC{2500}{\sphinxunichar{2500}}
  \sphinxDUC{2502}{\sphinxunichar{2502}}
  \sphinxDUC{2514}{\sphinxunichar{2514}}
  \sphinxDUC{251C}{\sphinxunichar{251C}}
  \sphinxDUC{2572}{\textbackslash}
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}



\usepackage{times}
\expandafter\ifx\csname T@LGR\endcsname\relax
\else
% LGR was declared as font encoding
  \substitutefont{LGR}{\rmdefault}{cmr}
  \substitutefont{LGR}{\sfdefault}{cmss}
  \substitutefont{LGR}{\ttdefault}{cmtt}
\fi
\expandafter\ifx\csname T@X2\endcsname\relax
  \expandafter\ifx\csname T@T2A\endcsname\relax
  \else
  % T2A was declared as font encoding
    \substitutefont{T2A}{\rmdefault}{cmr}
    \substitutefont{T2A}{\sfdefault}{cmss}
    \substitutefont{T2A}{\ttdefault}{cmtt}
  \fi
\else
% X2 was declared as font encoding
  \substitutefont{X2}{\rmdefault}{cmr}
  \substitutefont{X2}{\sfdefault}{cmss}
  \substitutefont{X2}{\ttdefault}{cmtt}
\fi


\usepackage[Sonny]{fncychap}
\ChNameVar{\Large\normalfont\sffamily}
\ChTitleVar{\Large\normalfont\sffamily}
\usepackage{sphinx}

\fvset{fontsize=\small}
\usepackage{geometry}


% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}


\usepackage{sphinxmessages}
\setcounter{tocdepth}{1}



\title{CCPM Biobank GWAS Pipeline}
\date{Jan 05, 2021}
\release{0.1.0}
\author{Tonya Brunetti}
\newcommand{\sphinxlogo}{\vbox{}}
\renewcommand{\releasename}{Release}
\makeindex
\begin{document}

\ifdefined\shorthandoff
  \ifnum\catcode`\=\string=\active\shorthandoff{=}\fi
  \ifnum\catcode`\"=\active\shorthandoff{"}\fi
\fi

\pagestyle{empty}
\sphinxmaketitle
\pagestyle{plain}
\sphinxtableofcontents
\pagestyle{normal}
\phantomsection\label{\detokenize{index::doc}}



\chapter{Installation}
\label{\detokenize{index:installation}}
Good news!  There is no installation required to use this pipeline and it should be OS agnostic.  There are only 2 true system dependencies and most HPCs and shared resources already have these dependencies installed:
\begin{itemize}
\item {} 
Singularity version \textgreater{}= 3.0

\item {} 
Golang version \textgreater{}= go1.13.5 (has been tested on versions go1.13.5 and go1.15.2 on a Linux OS).

\end{itemize}

If not, you can \sphinxhref{https://sylabs.io/docs/}{install Singularity} and \sphinxhref{https://golang.org/doc/install}{install Golang} using the links on your local computer.  Both should be available across OS platforms for Windows, MacOS, and Linux.


\chapter{Getting Started}
\label{\detokenize{index:getting-started}}

\section{Deciphering the config file}
\label{\detokenize{decipheringConfig:deciphering-the-config-file}}\label{\detokenize{decipheringConfig::doc}}
The config file is a text file that contains key\sphinxhyphen{}value mappings to all input parameters and pipeline logic. The file can be broken down into three main parts: the pipeline logic, environment setup, and the user data input options.


\subsection{Pipeline Logic}
\label{\detokenize{decipheringConfig:pipeline-logic}}
There are five keywords that control the logic of the pipeline and all five accept only boolean arguments (true/false):

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{GenerateGRM}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{GenerateNull}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{GenerateAssociations}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{GenerateResults}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{SkipChunking}\PYG{p}{:}\PYG{n}{false}
\end{sphinxVerbatim}


\subsection{Environment Setup}
\label{\detokenize{decipheringConfig:environment-setup}}
There are three keywords that control the environment setup of the pipeline. It is a requirement to use the container provided as an LSF object on github. This ensures all software is properly versioned and it reduces the complexity of the pipeline by using predefining software paths as well as reducing installation issues that arise with many softawre dependencies.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{BindPoint}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{bind}\PYG{o}{/}\PYG{n}{container}
\PYG{n}{BindPointTemp}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{tmp}\PYG{o}{/}
\PYG{n}{Container}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{SAIGE\PYGZus{}v0}\PYG{o}{.}\PYG{l+m+mi}{39}\PYG{n}{\PYGZus{}CCPM\PYGZus{}biobank\PYGZus{}singularity\PYGZus{}recipe\PYGZus{}file\PYGZus{}11162020}\PYG{o}{.}\PYG{n}{simg}
\end{sphinxVerbatim}

\sphinxcode{\sphinxupquote{BindPoint}} and \sphinxcode{\sphinxupquote{BindPointTemp}} are directories to where you want the container to be mounted.  \sphinxcode{\sphinxupquote{BindPointTemp}} allows you to give the container a secondary binding point where all temp files and calculations will be performed, and once finished will move the final files to the \sphinxcode{\sphinxupquote{BindPoint}}.  The files generated in \sphinxcode{\sphinxupquote{BindPointTemp}} will be deleted after the run completes.

\begin{sphinxadmonition}{note}{Note:}
If you do not want to use or do not have a temp directory, please set this parameter to be the same as \sphinxcode{\sphinxupquote{BindPoint}}.
\end{sphinxadmonition}

\begin{sphinxadmonition}{warning}{Warning:}
\sphinxstylestrong{IMPORTANT PLEASE READ!} The container follows the same rules of inheritance as Singularity specifies.  This means the \sphinxcode{\sphinxupquote{BindPoint}} and \sphinxcode{\sphinxupquote{BindPointTemp}} become the highest point in your directory tree.  Thus, you can only access paths, directories, and files if you are able to traverse below the tree from these starting points but not above these entry points from your host system.  Therefore, be sure all the paths, directories, and files specified in the config file are contained within the scope of at least one of these two entry points.
\end{sphinxadmonition}


\subsection{User Data Input}
\label{\detokenize{decipheringConfig:user-data-input}}
The remainder of the keywords are parameters offered to the user and the user can specify paths and options for all or some of these keywords:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{ChromosomeLengthFile}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{chromosomeLengths}\PYG{o}{.}\PYG{n}{txt}
\PYG{n}{Build}\PYG{p}{:}\PYG{n}{hg38}
\PYG{n}{Chromosomes}\PYG{p}{:}\PYG{l+m+mi}{1}\PYG{o}{\PYGZhy{}}\PYG{l+m+mi}{22}
\PYG{n}{ImputeSuffix}\PYG{p}{:}\PYG{n}{\PYGZus{}rsq70\PYGZus{}merged\PYGZus{}renamed}\PYG{o}{.}\PYG{n}{vcf}\PYG{o}{.}\PYG{n}{gz}
\PYG{n}{ImputeDir}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{imputed}\PYG{o}{/}\PYG{n}{data}\PYG{o}{/}\PYG{n}{directory}\PYG{o}{/}
\PYG{n}{OutDir}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{output}\PYG{o}{/}\PYG{n}{final}\PYG{o}{/}\PYG{n}{results}
\PYG{n}{OutPrefix}\PYG{p}{:}\PYG{n}{myGWAS}
\PYG{n}{PhenoFile}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{phenotype}\PYG{o}{/}\PYG{n}{covariate}\PYG{o}{/}\PYG{n}{file}
\PYG{n}{Plink}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{LDpruned}\PYG{o}{/}\PYG{n}{plink}\PYG{o}{/}\PYG{n}{file}\PYG{o}{/}\PYG{n}{prefix}
\PYG{n}{Trait}\PYG{p}{:}\PYG{n}{binary}
\PYG{n}{Pheno}\PYG{p}{:}\PYG{n}{myPhenotype}
\PYG{n}{InvNorm}\PYG{p}{:}\PYG{n}{FALSE}
\PYG{n}{Covars}\PYG{p}{:}\PYG{n}{PC1}\PYG{p}{,}\PYG{n}{PC2}\PYG{p}{,}\PYG{n}{PC3}\PYG{p}{,}\PYG{n}{PC4}\PYG{p}{,}\PYG{n}{PC5}\PYG{p}{,}\PYG{n}{age}\PYG{p}{,}\PYG{n}{sex}
\PYG{n}{SampleID}\PYG{p}{:}\PYG{n}{SampleIDcol}
\PYG{n}{NThreads}\PYG{p}{:}
\PYG{n}{SparseKin}\PYG{p}{:}\PYG{k+kc}{True}
\PYG{n}{Markers}\PYG{p}{:}\PYG{l+m+mi}{30}
\PYG{n}{Rel}\PYG{p}{:}\PYG{l+m+mf}{0.0625}
\PYG{n}{Loco}\PYG{p}{:}\PYG{n}{TRUE}
\PYG{n}{CovTransform}\PYG{p}{:}\PYG{k+kc}{True}
\PYG{n}{VcfField}\PYG{p}{:}\PYG{n}{DS}
\PYG{n}{MAF}\PYG{p}{:}\PYG{l+m+mf}{0.05}
\PYG{n}{MAC}\PYG{p}{:}\PYG{l+m+mi}{10}
\PYG{n}{IsDropMissingDosages}\PYG{p}{:}\PYG{n}{FALSE}
\PYG{n}{InfoFile}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{info}\PYG{o}{/}\PYG{n}{file}
\PYG{n}{SaveChunks}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{GrmMAF}\PYG{p}{:}\PYG{l+m+mf}{0.01}
\PYG{n}{ChunkVariants}\PYG{p}{:}\PYG{l+m+mi}{1000000}
\PYG{n}{SaveAsTar}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{ImputationFileList}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n+nb}{list}\PYG{o}{/}\PYG{n}{of}\PYG{o}{/}\PYG{n}{chunked}\PYG{o}{/}\PYG{n}{chromosomes}\PYG{o}{/}\PYG{n}{file}\PYG{o}{.}\PYG{n}{txt}
\PYG{n}{SparseGRM}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{grm}\PYG{o}{/}\PYG{n}{file}\PYG{o}{.}\PYG{n}{mtx}
\PYG{n}{SampleIDFile}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{grm}\PYG{o}{/}\PYG{n}{sample}\PYG{o}{/}\PYG{n+nb}{id}\PYG{o}{/}\PYG{n}{file}\PYG{o}{.}\PYG{n}{mtx}\PYG{o}{.}\PYG{n}{SampleID}\PYG{o}{.}\PYG{n}{txt}
\PYG{n}{NullModelFile}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{null}\PYG{o}{/}\PYG{n}{model}\PYG{o}{/}\PYG{n}{file}\PYG{o}{.}\PYG{n}{rda}
\PYG{n}{VarianceRatioFile}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{variance}\PYG{o}{/}\PYG{n}{ratio}\PYG{o}{/}\PYG{n}{file}\PYG{o}{.}\PYG{n}{txt}
\PYG{n}{AssociationFile}\PYG{p}{:}\PYG{o}{/}\PYG{n}{path}\PYG{o}{/}\PYG{n}{to}\PYG{o}{/}\PYG{n}{concatenated}\PYG{o}{/}\PYG{n}{association}\PYG{o}{/}\PYG{n}{results}\PYG{o}{/}\PYG{n}{file}\PYG{o}{.}\PYG{n}{txt}
\end{sphinxVerbatim}


\section{Parameters}
\label{\detokenize{parameters:parameters}}\label{\detokenize{parameters::doc}}
This explains in more detail how and when to use each keyword in the config file.  For readability, the minimum required parameters are listed for each step.  If more than one step will be run (\textgreater{}1 out of 5 possible pipeline logic keywords is set to true), the minimum parameters required is the sum of the minimum required parameters for all true keywords.


\subsection{Minimum Required Parameters for Each Step}
\label{\detokenize{parameters:minimum-required-parameters-for-each-step}}

\subsubsection{Minimum Requirements to Run Genetic Relatedness Matrix}
\label{\detokenize{grmParameters:minimum-requirements-to-run-genetic-relatedness-matrix}}\label{\detokenize{grmParameters::doc}}
To calculate the GRM, \sphinxcode{\sphinxupquote{GenerateGRM:true}} needs to be specified in the config file.  If it is set to \sphinxcode{\sphinxupquote{false}}, the pipeline assumes one of two things:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxstyleemphasis{GRM is not needed because one is provided in the config file from from a previous calculation}

\item {} 
\sphinxstyleemphasis{GRM is not needed at all for the scope of the pipeline logic}

\end{enumerate}


\subsubsection{Minimum Requirements to Run Null Model}
\label{\detokenize{nullModelParameters:minimum-requirements-to-run-null-model}}\label{\detokenize{nullModelParameters::doc}}
To calculate the null model, \sphinxcode{\sphinxupquote{GenerateNull:true}} needs to be specified in the config file.  If it is set to \sphinxcode{\sphinxupquote{false}}, the pipeline assumes one of two things:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxstyleemphasis{The null model files are not needed because one is provided in the config file from from a previous calculation}

\item {} 
\sphinxstyleemphasis{The null model files are not needed at all for the scope of the pipeline logic}

\end{enumerate}


\subsubsection{Minimum Requirements to Run Association Analysis}
\label{\detokenize{assocAnalysisParameters:minimum-requirements-to-run-association-analysis}}\label{\detokenize{assocAnalysisParameters::doc}}
To calculate the association analyis, \sphinxcode{\sphinxupquote{GenerateAssociations:true}} needs to be specified in the config file.  If it is set to \sphinxcode{\sphinxupquote{false}}, the pipeline assumes one of two things:
\begin{enumerate}
\sphinxsetlistlabels{\arabic}{enumi}{enumii}{}{.}%
\item {} 
\sphinxstyleemphasis{The association file results are not needed because one is provided in the config file from from a previous calculation}

\item {} 
\sphinxstyleemphasis{The association analysis files are not needed at all for the pipeline}

\end{enumerate}


\subsubsection{Minimum Requirements to Run Results}
\label{\detokenize{resultsParameters:minimum-requirements-to-run-results}}\label{\detokenize{resultsParameters::doc}}
To calculate the association analyis, \sphinxcode{\sphinxupquote{GenerateResults:true}} needs to be specified in the config file.  If it is set to \sphinxcode{\sphinxupquote{false}}, the pipeline assumes that no data clean up, data merges, and graphical summaries are needed.


\subsection{Full List of User Input Data Parameters}
\label{\detokenize{parameters:full-list-of-user-input-data-parameters}}

\begin{savenotes}\sphinxatlongtablestart\begin{longtable}[c]{|\X{20}{100}|\X{20}{100}|\X{10}{100}|\X{40}{100}|\X{10}{100}|}
\sphinxthelongtablecaptionisattop
\caption{Definitions and Use\strut}\label{\detokenize{parameters:id1}}\\*[\sphinxlongtablecapskipadjust]
\hline
\sphinxstyletheadfamily 
Parameter
&\sphinxstyletheadfamily 
Type
&\sphinxstyletheadfamily 
Default
&\sphinxstyletheadfamily 
Description
&\sphinxstyletheadfamily 
Relevance
\\
\hline
\endfirsthead

\multicolumn{5}{c}%
{\makebox[0pt]{\sphinxtablecontinued{\tablename\ \thetable{} \textendash{} continued from previous page}}}\\
\hline
\sphinxstyletheadfamily 
Parameter
&\sphinxstyletheadfamily 
Type
&\sphinxstyletheadfamily 
Default
&\sphinxstyletheadfamily 
Description
&\sphinxstyletheadfamily 
Relevance
\\
\hline
\endhead

\hline
\multicolumn{5}{r}{\makebox[0pt][r]{\sphinxtablecontinued{continues on next page}}}\\
\endfoot

\endlastfoot

ChromosomeLengthFile
&
string
&&
This is a tab\sphinxhyphen{}delimited text file that contains the same names as the chromosomes in your dataset followed by the chromosome length.  This file can generally be pulled down from the NCBI website under the build you are using (hg19 or hg38).  This is used for chunking the imputed data.
&
if \sphinxcode{\sphinxupquote{skipChunking:false}}
\\
\hline
Build
&
option: 
hg19 or hg38
&
hg38
&
This determines whether the software needs to be searching for a “chr” before the chomosome name or not
&\\
\hline
Chromosomes
&
string
&
1\sphinxhyphen{}22
&
This lets the software know which chromosomes you want to use for association analysis.  It must be a range.  If you want to only run analysis on a single chromsome, the start and end will be the same value.  For example:  running chromome 2 only will look like 2\sphinxhyphen{}2.
&
if \sphinxcode{\sphinxupquote{GenerateAssociations:true}}
\\
\hline
ImputeSuffix
&
string
&&
The full suffix for the software to determine which files in the directory are impuation files. Important, it is assuming the prefix is either a chromosome number (hg19) or the string chr followed by the chromosome number (hg38)
&
if \sphinxcode{\sphinxupquote{GenerateAssociations:true}} and \sphinxcode{\sphinxupquote{skipChunking:false}}
\\
\hline
ImputeDir
&
string
&&
Full path to directory where imputed results are located
&
if \sphinxcode{\sphinxupquote{GenerateAssociations:true}}
\\
\hline
OutDir
&
string
&&
Full path to directory where final results should be transferred
&
all
\\
\hline
OutPrefix
&
string
&&
string (no whitespace or special characters) to prefix to the output files generated
&
all
\\
\hline
PhenoFile
&
string
&&
Full path to tab\sphinxhyphen{}delimted phenotype file containing sample IDs, phenotypes, and covariates, with whatever string of  headers you choose. NO WHITESPACES in header names.
&
If :\sphinxcode{\sphinxupquote{GenerateNull:true}}
\\
\hline
Plink
&
string
&&
Full path to the directory and plink file prefix (dropping the suffix .bed,.bim,.fam) to an LD\sphinxhyphen{}pruned set of data to be used to generate GRM relatedness and to select random markers from for the variance ratio value
&
If \sphinxcode{\sphinxupquote{GenerateGRM:true}} or \sphinxcode{\sphinxupquote{GenerateNull:true}}
\\
\hline
Trait
&
option: binary or quantitative
&&
Based upon your association phenotype.  If binary, all values will be 0/1/NA, if quantitivate all phenotype traits to be analyzed will be continuous or have numeric quantitative meaning
&
if \sphinxcode{\sphinxupquote{GenerateNull:true}} or \sphinxcode{\sphinxupquote{GenerateResults:true}}
\\
\hline
Pheno
&
string
&&
The exact name (case\sphinxhyphen{}sensitive) of the phenotype to be analyzed in your PhenoFile.  Must be present in PhenoFile.
&
if \sphinxcode{\sphinxupquote{GenerateNull:true}} or \sphinxcode{\sphinxupquote{GenerateResults:true}}
\\
\hline
InvNorm
&
boolean
&
FALSE
&
This applies to the phenotype of interest to be analzyed and whether to perform an inverse normalization.  For binary traits, this should be set to FALSE and for quantitative traits, set this to TRUE.
&
if \sphinxcode{\sphinxupquote{GenerateNull:true}}
\\
\hline
Covars
&
comma\sphinxhyphen{}separated list
&&
A comma\sphinxhyphen{}separated list (no whitespaces) of all the covariate names to regress out in the model.  These variables need to be in your PhenoFile.
&
if \sphinxcode{\sphinxupquote{GenerateNull:true}} or \sphinxcode{\sphinxupquote{GenerateResults:true}}
\\
\hline
SampleID
&
string
&&
A string (no whitespaces) that is contained in the header of your PhenoFile.  This is the sampleID names and they must be the same names as listed in the PhenoFile, Imputation Files, and Plink Files.
&
if \sphinxcode{\sphinxupquote{GenerateNull:true}}
\\
\hline
Nthreads
&
int
&&
Strongly Recommned to leave this blank!  If left blank, it will auto\sphinxhyphen{}decect available resources and scale steps automatically on the back\sphinxhyphen{}end.  By specifiying the threads it tells the program to use max Nthreads for parallelization and concurrency.
&
all
\\
\hline
SparseKin
&
boolean
&
TRUE
&
If set to true, takes advantage of the sparsity of the GRM, otherwise will not use the sparsity to make assessments
&
if \sphinxcode{\sphinxupquote{GenerateNull:true}}
\\
\hline
Markers
&
int
&
30
&
The number of random markers selected from the LD\sphinxhyphen{}pruned plink file to estimate the variance ratio component in the null model. Warning!  This number increases time linearly
&
if \sphinxcode{\sphinxupquote{GenerateNull:true}}
\\
\hline
Rel
&
float
&
0.0625
&
A float between 0.0\sphinxhyphen{}1.0.  This is the threshold in kinship estimate to consider someone related.  Anything below this value will be treated as an unrelated individual in the pairwise comparison and calculation for the sparse GRM.
&
if \sphinxcode{\sphinxupquote{GenerateGRM:true}}
\\
\hline
Loco
&
boolean
&
TRUE
&
Leave\sphinxhyphen{}One\sphinxhyphen{}Chromosome\sphinxhyphen{}Out method.  Warning \textendash{} Setting this to true, increases the time complexity of the algorithm.
&
if \sphinxcode{\sphinxupquote{GenerateNull:true\textasciigrave{}and :code:\textasciigrave{}GenerateAssociations:true}}
\\
\hline
CovTransform
&
boolean
&
TRUE
&
Recommended to set to true.  It is a QR decomposition that aids in the covergence of the null model.
&
if \sphinxcode{\sphinxupquote{GenerateNull:true}}
\\
\hline
VcfField
&
option: DS or GT
&
DS
&
This determines what metric to base association upon.  DS = dosages and GT = genotypes.  If you have genotypes only, i.e. chip data withouth dosage calculations, DS cannot be used!
&
if \sphinxcode{\sphinxupquote{GenerateAssociations:true}}
\\
\hline
MAF
&
float
&
0.05
&
Float between 0.0\sphinxhyphen{}0.50 that specifies the cutoff to be considered a common snp or a rare snp.  For example, keeping this to the default of 0.05 will assume common snps are defined as those with a minor allele frequency \textgreater{}5\% and that rare snps are defined as those with a minor allele frequency ≤ 5\%.  THIS IS NOT A FILTER!
&
if \sphinxcode{\sphinxupquote{GenerateResults:true}}
\\
\hline
MAC
&
int
&
10
&
A filter applied to the cleaned association results to remove snps that have low minor allele counts.  Default recommendation is to set this to 10.
&
if \sphinxcode{\sphinxupquote{GenerateResults:true}}
\\
\hline
IsDropMissingDosages
&
boolean
&
FALSE
&&
if \sphinxcode{\sphinxupquote{GenerateAssociations:true}}
\\
\hline
InfoFile
&
string
&&
Path to the info file.  This file contains snps information pertaining to chromosome, positions, genotype/imputation status, R2, ER2 values. For formatting of this file please refer to \textless{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\sphinxhyphen{}\(\rightarrow\)
&
if \sphinxcode{\sphinxupquote{GenerateResults:true}}
\\
\hline
SaveChunks
&
boolean
&
TRUE
&
Specifies whether to save the chunked files and the queue list for future use.
&
if \sphinxcode{\sphinxupquote{GenerateAssociations:true}} and \sphinxcode{\sphinxupquote{skipChunking:false}}
\\
\hline
GrmMAF
&
float
&
0.01
&
The minor allele frequency threshold for a snp to be included in the GRM calculation based on the LD\sphinxhyphen{}pruned plink file.  For example, if set to 0.01 this means any snp with a MAF \textgreater{} 0.01 wil be used to calculate relatedness in the GRM.
&
if \sphinxcode{\sphinxupquote{GenerateGRM:true}}
\\
\hline
ChunkVariants
&
int
&
1000000
&&
if \sphinxcode{\sphinxupquote{GenerateNull:true}} and \sphinxcode{\sphinxupquote{SkipChunking:false}}
\\
\hline
SaveAsTar
&
boolean
&
FALSE
&&
all
\\
\hline
ImputationFileList
&
string
&&
Ends in \_chunkedImputationQueue.txt
&
if \sphinxcode{\sphinxupquote{GenerateAssociations:true}} and \sphinxcode{\sphinxupquote{skipChunking:true}}
\\
\hline
SparseGRM
&
string
&&
Ends in .sparseGRM.mtx
&
if \sphinxcode{\sphinxupquote{GenerateGRM:false}} and \sphinxcode{\sphinxupquote{GenerateNull:true}}
\\
\hline
SampleIDFile
&
string
&&
Ends in sparseGRM.mtx.sampleIDs.txt
&
if \sphinxcode{\sphinxupquote{GenerateGRM:false}} and \sphinxcode{\sphinxupquote{GenerateNull:true}}
\\
\hline
NullModelFile
&
string
&&
Ends in .rda
&
if \sphinxcode{\sphinxupquote{GenerateNull:false}} and \sphinxcode{\sphinxupquote{GenerateAssociations:true}}
\\
\hline
VarianceRatioFile
&
string
&&
Ends in .varianceRatio.txt
&
if \sphinxcode{\sphinxupquote{GenerateNull:false}} and \sphinxcode{\sphinxupquote{GenerateAssociations:true}}
\\
\hline
AssociationFile
&
string
&&
Ends in \_SNPassociationAnalysis.txt
&
If \sphinxcode{\sphinxupquote{GenerateAssociations:false}} and \sphinxcode{\sphinxupquote{GenerateResults:true}}
\\
\hline
\end{longtable}\sphinxatlongtableend\end{savenotes}


\section{Formatting the Required Files}
\label{\detokenize{fileFormats:formatting-the-required-files}}\label{\detokenize{fileFormats::doc}}
The pipeline does require some formatting for file names and contents within files to be present.  This section explains the file format expectations, as well as file name expectations.


\subsection{Parameter: ChromosomeLengthFile}
\label{\detokenize{fileFormats:parameter-chromosomelengthfile}}
The \sphinxcode{\sphinxupquote{ChromsomeLengthFile}} parameter is a file that can be downloaded and modified from sources such as \sphinxhref{https://www.ncbi.nlm.nih.gov/grc/human/data/}{NCBI} and \sphinxhref{http://hgdownload.cse.ucsc.edu/goldenpath/hg38/bigZips/hg38.chrom.sizes}{UCSC}.  It should be a tab\sphinxhyphen{}delimted file with chromosomes 1\sphinxhyphen{}22 listed in the first column and the length of the chromosome in the second column. This file contains \sphinxstylestrong{no header}.  The build you select should be based upon the build used for your imputation file.  If only using genotyping file, base this file on the genotype build.

\noindent\sphinxincludegraphics[width=400\sphinxpxdimen]{{chromosomeLengthFile}.png}

\begin{sphinxShadowBox}
\sphinxstylesidebartitle{\sphinxstylestrong{Example snippet for hg38}}
\begin{itemize}
\item {} 
1st column is chromosome

\item {} 
2nd column is length

\end{itemize}
\end{sphinxShadowBox}

\begin{sphinxadmonition}{warning}{Warning:}
\sphinxstylestrong{hg19 vs hg38 formatting}  Depending on the build you choose you must format your file accordingly.
\begin{itemize}
\item {} 
\sphinxstyleemphasis{hg19:} omit "chr" and should strictly be integer values between 1\sphinxhyphen{}22 for the first column.

\item {} 
\sphinxstyleemphasis{hg38:} include the preceding "chr" string (no spaces) for integer values between 1\sphinxhyphen{}22 for the first column.

\end{itemize}
\end{sphinxadmonition}


\subsection{Parameter: ImputeSuffix}
\label{\detokenize{fileFormats:parameter-imputesuffix}}
This string is the suffix of your imputation file.  The imputation files require to be named and split in a particular way.
\begin{description}
\item[{First, all imputation files follow these rules:}] \leavevmode\begin{itemize}
\item {} 
split by chromosome

\item {} 
sample names are the samples names used in \sphinxcode{\sphinxupquote{GenerateGRM}} and \sphinxcode{\sphinxupquote{GenerateNull}} (order doesn\textquotesingle{}t need to be maintained)

\item {} 
gzipped vcf files, and

\item {} 
have the matching tabix index ending in the imputation file name followed by the .tbi string

\end{itemize}

\end{description}

This is how each imputation file needs to be named:

\sphinxstylestrong{for hg38}

\noindent\sphinxincludegraphics[width=800\sphinxpxdimen]{{imputationSuffix_hg38}.png}

\sphinxstylestrong{for hg19}

\noindent\sphinxincludegraphics[width=800\sphinxpxdimen]{{imputationSuffix_hg19}.png}

\begin{sphinxadmonition}{warning}{Warning:}
\sphinxstylestrong{hg19 vs hg38 formatting}  Depending on the build you choose you must format your file name accordingly.
\begin{itemize}
\item {} 
\sphinxstyleemphasis{hg19:} omit "chr" and strictly be integer values between 1\sphinxhyphen{}22 as the prefix, followed by a required underscore, following by remaining string.

\item {} 
\sphinxstyleemphasis{hg38:} include the preceding "chr" string (no spaces) for integer values between 1\sphinxhyphen{}22 as the prefix, followed by a required underscore, following by remaining string.

\end{itemize}
\end{sphinxadmonition}


\subsection{Parameter: PhenoFile}
\label{\detokenize{fileFormats:parameter-phenofile}}
This is tab\sphinxhyphen{}delimited txt file that contains all the sample IDs (must be the same IDs used in the plink file, imputation file, GRM, and null model \sphinxhyphen{}\sphinxhyphen{} order agnostic).  In addition to the sample IDs, it also contains any phenotype(s) you may want to run and any covariates you may want to use, although the user is not required to use everything listed in the header/file.

Below is an example of a tab\sphinxhyphen{}delimited \sphinxcode{\sphinxupquote{PhenoFile}}.  Again, none of these header names are required, however, there needs to be a header variable at minimum that denotes the sample ID and a phenotype to analyze.

\noindent\sphinxincludegraphics[width=800\sphinxpxdimen]{{phenoFile}.png}

\begin{sphinxadmonition}{note}{Note:}
It is worthwhile to generate a single PhenoFile that contains many phenotypes and covariates you may want to analyze for the sample set.  Within the config file, the user can specify which single phenotype to run and which covariates to run.  That way, the user can run several jobs in parallel using the same PhenoFile but just changing the trait, phenotype, covariates, and invNorm parameters without having to change anything else.
\end{sphinxadmonition}


\subsection{Parameter: Plink}
\label{\detokenize{fileFormats:parameter-plink}}

\subsection{Parameter: Pheno}
\label{\detokenize{fileFormats:parameter-pheno}}

\subsection{Parameter: Covars}
\label{\detokenize{fileFormats:parameter-covars}}

\subsection{Parameter: SampleID}
\label{\detokenize{fileFormats:parameter-sampleid}}

\subsection{Parameter: InfoFile}
\label{\detokenize{fileFormats:parameter-infofile}}

\subsection{Parameter: ImputationFileList}
\label{\detokenize{fileFormats:parameter-imputationfilelist}}

\subsection{Parameter: SparseGRM}
\label{\detokenize{fileFormats:parameter-sparsegrm}}

\subsection{Parameter: SampleIDFile}
\label{\detokenize{fileFormats:parameter-sampleidfile}}

\subsection{Parameter: NullModelFile}
\label{\detokenize{fileFormats:parameter-nullmodelfile}}

\subsection{Parameter: VarianceRatioFile}
\label{\detokenize{fileFormats:parameter-varianceratiofile}}

\subsection{Parameter: AssociationFile}
\label{\detokenize{fileFormats:parameter-associationfile}}

\section{Output Generated}
\label{\detokenize{output:output-generated}}\label{\detokenize{output::doc}}

\section{Parsing Through StdErr and StdOut}
\label{\detokenize{parsingStdErrOut:parsing-through-stderr-and-stdout}}\label{\detokenize{parsingStdErrOut::doc}}
When using the pipeline with the container and go binary file, all standard out messages and errors are reported in the generated log file.  It is important to note, that since the pipeline automatically runs steps concurrently when possible, the order of the log files may show some steps running as dependencies become available and based on the threads available.  Also, this standard out is used by SAIGE as well, so anything that does not follow the format below is a results of the software calculations generated by SAIGE, and not the pipeline.

The following shows how messages are formatted in the log output:

\noindent\sphinxincludegraphics[width=800\sphinxpxdimen]{{messageFormat}.png}


\subsection{Logging Levels}
\label{\detokenize{parsingStdErrOut:logging-levels}}
There are 4 different levels or types of debugging and error handling.  Some are meant to inform, while other are meant to stop the pipeline if an error occurs that cannot be resolved.


\begin{savenotes}\sphinxattablestart
\centering
\begin{tabulary}{\linewidth}[t]{|T|T|}
\hline
\sphinxstyletheadfamily 
Level
&\sphinxstyletheadfamily 
Interpretation
\\
\hline
UPDATE!
&
To inform the user of the status of the pipeline.
\\
\hline
CONFIRMED!
&
To let user know a check or validation has been passed.
\\
\hline
WARNING!
&
There may be a problem but it continues through the pipeline and skips the issue.
\\
\hline
ERROR!
&
There is definitely a problem and forces the pipeline to prematurely exit.
\\
\hline
\end{tabulary}
\par
\sphinxattableend\end{savenotes}

By using the levels and function names, one can \sphinxcode{\sphinxupquote{grep}} through the log file to find specific messages related to functions or error types.


\subsection{Exit Status Codes}
\label{\detokenize{parsingStdErrOut:exit-status-codes}}
Although I am not a huge fan of relying on exit status codes, for those that use Gnu Parallel or other software that records exit status to determine if a job was successful or not, I have coded up a few exit codes so that if an error does occur the number can tell the user which function is responsible for the error.  This make troubleshooting and errors easier to identify.


\begin{savenotes}\sphinxattablestart
\centering
\begin{tabulary}{\linewidth}[t]{|T|T|}
\hline
\sphinxstyletheadfamily 
functions
&\sphinxstyletheadfamily 
Exit Status
\\
\hline
func(main)
&
return 42
\\
\hline
func(chunk)
&
return 10
\\
\hline
func(smallerChunk)
&
None
\\
\hline
func(processing)
&
None
\\
\hline
func(nullModel)
&
None
\\
\hline
func(associationAnalysis)
&
None
\\
\hline
func(checkInput)
&
return 5
\\
\hline
func(saveResults)
&
return 99
\\
\hline
func(saveQueue)
&
None
\\
\hline
func(usePrevChunks)
&
return 17
\\
\hline
func(findElement)
&
return 20
\\
\hline
func(parser)
&
return 3
\\
\hline
\end{tabulary}
\par
\sphinxattableend\end{savenotes}


\chapter{Quick Start and Examples}
\label{\detokenize{index:quick-start-and-examples}}

\section{Example Work Flows}
\label{\detokenize{exampleWorkFlows:example-work-flows}}\label{\detokenize{exampleWorkFlows::doc}}

\subsection{Quick Start Command}
\label{\detokenize{exampleWorkFlows:quick-start-command}}
In order to run the pipeline, open a shell or bash prompt (or batch script for a job\sphinxhyphen{}scheduler) and type:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYGZdl{} ./CCPM\PYGZus{}GWAS\PYGZus{}pipeline myConfigFile.txt
\end{sphinxVerbatim}


\subsection{Full Pipeline}
\label{\detokenize{exampleWorkFlows:full-pipeline}}
\sphinxstylestrong{Full pipline} means you want to run every component of the pipeline from beginning to end in one go, without re\sphinxhyphen{}using any previously calculated data from the pipeline.  This is analagous to setting the pipeline logic kewords to the following:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{GenerateGRM}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{GenerateNull}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{GenerateAssociations}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{GenerateResults}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{SkipChunking}\PYG{p}{:}\PYG{n}{false}
\end{sphinxVerbatim}


\subsection{GRM only}
\label{\detokenize{exampleWorkFlows:grm-only}}
\sphinxstylestrong{GRM only} means you want to run the GRM step.  This is analagous to setting the pipeline logic kewords to the following:

\sphinxstyleemphasis{Note,} \sphinxcode{\sphinxupquote{SkipChunking}} \sphinxstyleemphasis{can be set to either} \sphinxcode{\sphinxupquote{true}} \sphinxstyleemphasis{or} \sphinxcode{\sphinxupquote{false}} \sphinxstyleemphasis{because it is only used if} \sphinxcode{\sphinxupquote{GenerateAssociation}} \sphinxstyleemphasis{is set to} \sphinxcode{\sphinxupquote{true}}.

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{GenerateGRM}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{GenerateNull}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{GenerateAssociations}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{GenerateResults}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{SkipChunking}\PYG{p}{:}\PYG{n}{false}
\end{sphinxVerbatim}


\subsection{Null Model only}
\label{\detokenize{exampleWorkFlows:null-model-only}}
\sphinxstylestrong{Null Model only} means you want to run the Null Model only.  It makes an assumption that you already have the GRM pre\sphinxhyphen{}calculated and want to re\sphinxhyphen{}use it in this step by setting the keywords \sphinxcode{\sphinxupquote{SparseGRM}} and \sphinxcode{\sphinxupquote{SampleIDFile}} located in the config file.  These two files are the result of running \sphinxcode{\sphinxupquote{GenerateGRM:true}}.

Choosing to run just the null model generation step is analagous to setting the pipeline logic kewords to the following:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{GenerateGRM}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{GenerateNull}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{GenerateAssociations}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{GenerateResults}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{SkipChunking}\PYG{p}{:}\PYG{n}{false}
\end{sphinxVerbatim}

\sphinxcode{\sphinxupquote{SkipChunking}} \sphinxstyleemphasis{can be set to either} \sphinxcode{\sphinxupquote{true}} \sphinxstyleemphasis{or} \sphinxcode{\sphinxupquote{false}} \sphinxstyleemphasis{because it is only used if} \sphinxcode{\sphinxupquote{GenerateAssociation}} \sphinxstyleemphasis{is set to} \sphinxcode{\sphinxupquote{true}}.


\subsection{Association Analyses Only}
\label{\detokenize{exampleWorkFlows:association-analyses-only}}
\sphinxstylestrong{Association Analyses only} means you only want to run the association anlysis.  It makes an assumption that you already have the null model file (.rda) pre\sphinxhyphen{}calculated and have a pre\sphinxhyphen{}calculate variance ratio file (.varianceRatio.txt) and want to re\sphinxhyphen{}use/use it in this step by setting the keywords \sphinxcode{\sphinxupquote{NullModelFile}} and \sphinxcode{\sphinxupquote{VarianceRatioFile}} located in the config file.  These two files are the result of running \sphinxcode{\sphinxupquote{GenerateNull:true}}.

Choosing to run just the association analysis step is analagous to setting the pipeline logic kewords to the following:

\begin{sphinxVerbatim}[commandchars=\\\{\}]
\PYG{n}{GenerateGRM}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{GenerateNull}\PYG{p}{:}\PYG{n}{false}
\PYG{n}{GenerateAssociations}\PYG{p}{:}\PYG{n}{true}
\PYG{n}{GenerateResults}\PYG{p}{:}\PYG{n}{false}
\end{sphinxVerbatim}

When \sphinxcode{\sphinxupquote{GenerateAssociations:true}}, the \sphinxcode{\sphinxupquote{SkipChunking}} logic comes into play.

\begin{sphinxadmonition}{note}{Note:}
This step produces the raw associaions results concatenated into a file.  It does not clean up the data, perform the proper flips, or generate graphs/figures.  If you want the raw data in addition to the previously mentioned actions, be sure to also set \sphinxcode{\sphinxupquote{GenerateResults:true}}.
\end{sphinxadmonition}


\subsection{Results and Graphs Only}
\label{\detokenize{exampleWorkFlows:results-and-graphs-only}}
\sphinxstylestrong{Results and Graphs Only} cleans up raw data that was previously generated from an association analysis and generates cleaned data in addition to some figures/graphs.  You can use any association analysis here as long as it meets the file formatting specifications for supplying results in \sphinxcode{\sphinxupquote{AssociationFile}}.


\subsection{Reuse Previously Indexed and Chunked Files}
\label{\detokenize{exampleWorkFlows:reuse-previously-indexed-and-chunked-files}}

\subsection{Combinations of Logic}
\label{\detokenize{exampleWorkFlows:combinations-of-logic}}

\chapter{FAQs}
\label{\detokenize{index:faqs}}

\chapter{Acknowledgements}
\label{\detokenize{index:acknowledgements}}

\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}



\renewcommand{\indexname}{Index}
\printindex
\end{document}