
<!DOCTYPE html>

<html lang="ls">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Output Generated &#8212; SAIGE-BRUSH v0.39 documentation</title>
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Parsing Through StdErr and StdOut" href="parsingStdErrOut.html" />
    <link rel="prev" title="Minimum Requirements to Run Results" href="resultsParameters.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="parsingStdErrOut.html" title="Parsing Through StdErr and StdOut"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="resultsParameters.html" title="Minimum Requirements to Run Results"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">SAIGE-BRUSH v0.39 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Output Generated</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="output-generated">
<h1>Output Generated<a class="headerlink" href="#output-generated" title="Permalink to this headline">¶</a></h1>
<p>This summarizes the expected output depending on how the pipeline logic boolean values are set in the config file.</p>
<div class="section" id="full-pipeline-with-chunking">
<h2>Full Pipeline with Chunking<a class="headerlink" href="#full-pipeline-with-chunking" title="Permalink to this headline">¶</a></h2>
<p>Setting this logic:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GenerateGRM</span><span class="p">:</span><span class="n">true</span>
<span class="n">GenerateNull</span><span class="p">:</span><span class="n">true</span>
<span class="n">GenerateAssociations</span><span class="p">:</span><span class="n">true</span>
<span class="n">GenerateResults</span><span class="p">:</span><span class="n">true</span>
<span class="n">SkipChunking</span><span class="p">:</span><span class="n">false</span>
<span class="n">SaveChunks</span><span class="p">:</span><span class="n">true</span>
</pre></div>
</div>
<p>Returns this output:</p>
<a class="reference internal image-reference" href="_images/fullPipeline_output.png"><img alt="_images/fullPipeline_output.png" class="align-center" src="_images/fullPipeline_output.png" style="width: 1000px;" /></a>
<p>What can be reused?</p>
</div>
<div class="section" id="genetic-relatedness-matrix-grm-only">
<h2>Genetic Relatedness Matrix (GRM) only<a class="headerlink" href="#genetic-relatedness-matrix-grm-only" title="Permalink to this headline">¶</a></h2>
<p>Setting this logic:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GenerateGRM</span><span class="p">:</span><span class="n">true</span>
<span class="n">GenerateNull</span><span class="p">:</span><span class="n">false</span>
<span class="n">GenerateAssociations</span><span class="p">:</span><span class="n">false</span>
<span class="n">GenerateResults</span><span class="p">:</span><span class="n">false</span>
<span class="n">SkipChunking</span><span class="p">:</span><span class="n">false</span>
</pre></div>
</div>
<p>Returns this output:</p>
<a class="reference internal image-reference" href="_images/grmOnly_output.png"><img alt="_images/grmOnly_output.png" class="align-center" src="_images/grmOnly_output.png" style="width: 1000px;" /></a>
<p>What can be reused?</p>
<a class="reference internal image-reference" href="_images/grmOnly_output_results_nextSteps.png"><img alt="_images/grmOnly_output_results_nextSteps.png" class="align-center" src="_images/grmOnly_output_results_nextSteps.png" style="width: 1000px;" /></a>
</div>
<div class="section" id="null-model-only">
<h2>Null Model only<a class="headerlink" href="#null-model-only" title="Permalink to this headline">¶</a></h2>
<p>Setting this logic:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GenerateGRM</span><span class="p">:</span><span class="n">false</span>
<span class="n">GenerateNull</span><span class="p">:</span><span class="n">true</span>
<span class="n">GenerateAssociations</span><span class="p">:</span><span class="n">false</span>
<span class="n">GenerateResults</span><span class="p">:</span><span class="n">false</span>
<span class="n">SkipChunking</span><span class="p">:</span><span class="n">false</span>
</pre></div>
</div>
<p>Returns this output:</p>
<a class="reference internal image-reference" href="_images/nullOnly_output.png"><img alt="_images/nullOnly_output.png" class="align-center" src="_images/nullOnly_output.png" style="width: 1000px;" /></a>
<p>What can be reused?</p>
<a class="reference internal image-reference" href="_images/nullOnly_output_results_nextSteps.png"><img alt="_images/nullOnly_output_results_nextSteps.png" class="align-center" src="_images/nullOnly_output_results_nextSteps.png" style="width: 1000px;" /></a>
</div>
<div class="section" id="association-analysis-only">
<h2>Association Analysis only<a class="headerlink" href="#association-analysis-only" title="Permalink to this headline">¶</a></h2>
<p>Setting this logic (see <a class="reference internal" href="generateAssociationOnlyTutorial.html#option-a"><span class="std std-ref">OPTION A: Imputation files need window chunking</span></a>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GenerateGRM</span><span class="p">:</span><span class="n">false</span>
<span class="n">GenerateNull</span><span class="p">:</span><span class="n">false</span>
<span class="n">GenerateAssociations</span><span class="p">:</span><span class="n">true</span>
<span class="n">GenerateResults</span><span class="p">:</span><span class="n">false</span>
<span class="n">SkipChunking</span><span class="p">:</span><span class="n">false</span>
<span class="n">SaveChunks</span><span class="p">:</span><span class="n">true</span>
</pre></div>
</div>
<p>Returns this output:</p>
<p>What can be reused?</p>
<blockquote>
<div><p>Setting this logic (see <a class="reference internal" href="generateAssociationOnlyTutorial.html#option-b"><span class="std std-ref">OPTION B: Imputation files do not need window chunking and will be reused</span></a>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">GenerateGRM</span><span class="p">:</span><span class="n">false</span>
<span class="n">GenerateNull</span><span class="p">:</span><span class="n">false</span>
<span class="n">GenerateAssociations</span><span class="p">:</span><span class="n">true</span>
<span class="n">GenerateResults</span><span class="p">:</span><span class="n">false</span>
<span class="n">SkipChunking</span><span class="p">:</span><span class="n">true</span>
</pre></div>
</div>
</div></blockquote>
<p>Returns this output:</p>
<p>What can be reused?</p>
</div>
</div>
<div class="section" id="output-files-generated">
<h1>Output Files Generated<a class="headerlink" href="#output-files-generated" title="Permalink to this headline">¶</a></h1>
<p>This section describes the file format output and what is located in each result file.</p>
<p>For association analysis results file per snp.  This is for text files generated with the suffix:
* *_allSNPs_noFiltering.txt.gz
* *_commonSNPs_cleaned.txt.gz
* *_rareSNPs_cleaned.txt.gz</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 14%" />
<col style="width: 86%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column</p></th>
<th class="head"><p>Interpretation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CHR</p></td>
<td><p>chromosome the variant is located</p></td>
</tr>
<tr class="row-odd"><td><p>POS</p></td>
<td><p>base position of the variant (hg19 vs hg38 is based upon what was specified at runtime)</p></td>
</tr>
<tr class="row-even"><td><p>majorAllele</p></td>
<td><p>major allele (most common frequency in the input population)</p></td>
</tr>
<tr class="row-odd"><td><p>minorAllele</p></td>
<td><p>minor allele (lesser common frequency in the input population)</p></td>
</tr>
<tr class="row-even"><td><p>SNPID</p></td>
<td><p>variant name</p></td>
</tr>
<tr class="row-odd"><td><p>BETA</p></td>
<td><p>effect size based on the minor allele</p></td>
</tr>
<tr class="row-even"><td><p>SE</p></td>
<td><p>standard error of beta</p></td>
</tr>
<tr class="row-odd"><td><p>OR</p></td>
<td><p>odds Ratio</p></td>
</tr>
<tr class="row-even"><td><p>LogOR</p></td>
<td><p>log(odds ratio)</p></td>
</tr>
<tr class="row-odd"><td><p>Lower95OR</p></td>
<td><p>lower 95% confidence interval of the odds ratio</p></td>
</tr>
<tr class="row-even"><td><p>Upper95OR</p></td>
<td><p>upper 95% confidence interval of the odds ratio</p></td>
</tr>
<tr class="row-odd"><td><p>MAF</p></td>
<td><p>minor allele frequency</p></td>
</tr>
<tr class="row-even"><td><p>MAC</p></td>
<td><p>minor allele count</p></td>
</tr>
<tr class="row-odd"><td><p>p.value</p></td>
<td><p>p-value of the variant</p></td>
</tr>
<tr class="row-even"><td><p>N.Cases</p></td>
<td><p>number of cases evaluated for this variant (binary traits only)</p></td>
</tr>
<tr class="row-odd"><td><p>N.Controls</p></td>
<td><p>number of controls evaluated for this variant (binary traits only)</p></td>
</tr>
<tr class="row-even"><td><p>casesHomMinor</p></td>
<td><p>number of cases evaluated for this variant that are homozygous for the minor allele (binary traits only)</p></td>
</tr>
<tr class="row-odd"><td><p>casesHet</p></td>
<td><p>number of cases evaluated for this variant that are heterozygous (binary traits only)</p></td>
</tr>
<tr class="row-even"><td><p>controlHomMinor</p></td>
<td><p>number of control evaluated for this variant that are homozygous for the minor allele (binary traits only)</p></td>
</tr>
<tr class="row-odd"><td><p>controlHet</p></td>
<td><p>number of controls evaluated for this variant that are heterozygous (binary traits only)</p></td>
</tr>
<tr class="row-even"><td><p>negLog10pvalue</p></td>
<td><p>negative log base 10 of the p-value</p></td>
</tr>
<tr class="row-odd"><td><p>R2</p></td>
<td><p>imputation quality for imputed variants</p></td>
</tr>
<tr class="row-even"><td><p>ER2</p></td>
<td><p>empirical R2 quality extracted from info file from imputation -- for genoytyped variants only</p></td>
</tr>
<tr class="row-odd"><td><p>GENOTYPE_STATUS</p></td>
<td><p>imputed, genotyped or genotyped and imputed status for the variant</p></td>
</tr>
</tbody>
</table>
<p>There is an intermediate file also output with the suffix <code class="code docutils literal notranslate"><span class="pre">allChromosomeResultsMerged.txt</span></code>.  This is output directly from SAIGE,
however, it is not cleaned and alleles and betas have not been corrected for minor allele basis.  All GWAS results, including the graphs
generated from the pipeline are based on the files above ending in <code class="code docutils literal notranslate"><span class="pre">_commonSNPs_cleaned.txt.gz</span></code> and <code class="code docutils literal notranslate"><span class="pre">_rareSNPs_cleaned.txt.gz</span></code>.</p>
<p>However, some T statistics and variance calculations are included in this file, along if the variant converged or not.  Therefore, it is still
saved in the output so it can easily be searched and referenced for particular variants, if needed.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 80%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column</p></th>
<th class="head"><p>Interpretation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>CHR</p></td>
<td><p>chromosome the variant is located</p></td>
</tr>
<tr class="row-odd"><td><p>POS</p></td>
<td><p>base position of the variant</p></td>
</tr>
<tr class="row-even"><td><p>SNPID</p></td>
<td><p>variant name</p></td>
</tr>
<tr class="row-odd"><td><p>Allele1</p></td>
<td><p>allele 1 (careful, not necessarily major allele)</p></td>
</tr>
<tr class="row-even"><td><p>Allele2</p></td>
<td><p>allele 2 (careful, not necessarily minor allele)</p></td>
</tr>
<tr class="row-odd"><td><p>AC_Allele2</p></td>
<td><p>allele count of allele 2</p></td>
</tr>
<tr class="row-even"><td><p>AF_Allele2</p></td>
<td><p>allele frequency of allele 2 (not necessarily minor allele)</p></td>
</tr>
<tr class="row-odd"><td><p>imputationInfo</p></td>
<td><p>imputation quality score</p></td>
</tr>
<tr class="row-even"><td><p>N</p></td>
<td><p>total samples considered in model for this variant</p></td>
</tr>
<tr class="row-odd"><td><p>BETA</p></td>
<td><p>effect size of allele 2 (not necessarily minor allele)</p></td>
</tr>
<tr class="row-even"><td><p>SE</p></td>
<td><p>standard error of beta</p></td>
</tr>
<tr class="row-odd"><td><p>p.value</p></td>
<td><p>p-value of the variant</p></td>
</tr>
<tr class="row-even"><td><p>p.value.NA</p></td>
<td><p>p-value of variant when SPA is not applied</p></td>
</tr>
<tr class="row-odd"><td><p>Is.SPA.converge</p></td>
<td><p>whether saddle point approximation (SPA) converged for this variant</p></td>
</tr>
<tr class="row-even"><td><p>Tstat</p></td>
<td><p>estimated variance of score statistic with sample related incorporated</p></td>
</tr>
<tr class="row-odd"><td><p>varTstar</p></td>
<td><p>variance of score statistic without sample related incorporated</p></td>
</tr>
<tr class="row-even"><td><p>AF.Cases</p></td>
<td><p>allele frequency of allele 2 in cases (only for binary trait)</p></td>
</tr>
<tr class="row-odd"><td><p>AF.Controls</p></td>
<td><p>allele frequency of allele 2 in controls (only for binary traits)</p></td>
</tr>
<tr class="row-even"><td><p>N.Cases</p></td>
<td><p>total cases considered in model for this variant (only for binary traits)</p></td>
</tr>
<tr class="row-odd"><td><p>N.Controls</p></td>
<td><p>total controls considered in model for this variant (only for binary traits)</p></td>
</tr>
<tr class="row-even"><td><p>homN_Allele2_cases</p></td>
<td><p>homozygote counts in cases</p></td>
</tr>
<tr class="row-odd"><td><p>hetN_Allele2_cases</p></td>
<td><p>heterozygote counts in cases</p></td>
</tr>
<tr class="row-even"><td><p>homN_Allele2_ctrls</p></td>
<td><p>homozygote counts in controls</p></td>
</tr>
<tr class="row-odd"><td><p>hetN_Allele2_ctrls</p></td>
<td><p>heterozygote counts in controls</p></td>
</tr>
</tbody>
</table>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Output Generated</a><ul>
<li><a class="reference internal" href="#full-pipeline-with-chunking">Full Pipeline with Chunking</a></li>
<li><a class="reference internal" href="#genetic-relatedness-matrix-grm-only">Genetic Relatedness Matrix (GRM) only</a></li>
<li><a class="reference internal" href="#null-model-only">Null Model only</a></li>
<li><a class="reference internal" href="#association-analysis-only">Association Analysis only</a></li>
</ul>
</li>
<li><a class="reference internal" href="#output-files-generated">Output Files Generated</a></li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="resultsParameters.html"
                        title="previous chapter">Minimum Requirements to Run Results</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="parsingStdErrOut.html"
                        title="next chapter">Parsing Through StdErr and StdOut</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/output.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="parsingStdErrOut.html" title="Parsing Through StdErr and StdOut"
             >next</a> |</li>
        <li class="right" >
          <a href="resultsParameters.html" title="Minimum Requirements to Run Results"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">SAIGE-BRUSH v0.39 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Output Generated</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, Tonya Brunetti.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.5.4.
    </div>
  </body>
</html>